### DISCUSSION TITLE
Feature wishlist

### DISCUSSION THREAD

**Discussion by chrysn** [ariel-os/ariel-os#642] [General]
Exceeding what we have concrete issues *and plans* for, this list is for collecting desirable features that we're still gauging interest in.

If you have features you'd like to see -- anything from integrated software down to MCU and board support, please add that to this list – and generously assign thumbs-up you would like to see implemented.

**Comment by emmanuelsearch** [comment]
[embedded-graphics](https://github.com/embedded-graphics/embedded-graphics) support!

**Reply by chrysn** [reply]
Is there hardware we want supported to go with that, or a concrete demo application you have in mind?

**Reply by chrysn** [reply]
Hardware that apparently we already run on is the [Ulanzi TC001](https://www.ulanzi.com/products/ulanzi-pixel-smart-clock-2882). Others have recommended that we look into what is apparently known as The Cheap Yellow Display. The [fossasia badge](https://github.com/RIOT-OS/RIOT/issues/20820) is also a good candidate.

**Reply by chrysn** [reply]
By now we have [the CDC badge](https://github.com/riatlabs/cdc-badge) somewhere [on the horizon](https://floss.social/@ariel/116000687022614910) with e-paper.

Already supported is the [STM32U083C SDC](https://ariel-os.github.io/ariel-os/dev/docs/book/boards/stm32u083c-dk.html), but as that has a segment LCD, it likely won't go through embedded-graphics (but is still something to have in mind w/rt visual output).

**Reply by chrysn** [reply]
Now tracking a concrete plan (without any timeline) in https://github.com/ariel-os/ariel-os/issues/1947

**Comment by chrysn** [comment]
Wireguard suport: This has been requested in this year's RustWeek Q&A.

Frankly, I always prefer to just have proper application-layer security on an untrusted and capable Internet, but if eg. the Internet connection is not capable (hello FH network without IPv6), this might be neat.

There is the [onetun crate](https://crates.io/crates/onetun) which at a brief glance appears to have all the parts. (The requester did imply that there are implementations for smoltcp, and smoltcp is mentioned in there).

**Reply by SimonIT** [reply]
By the way, I'm currently trying to do this here https://github.com/SimonIT/ariel-os-wireguard with an example https://github.com/SimonIT/ariel-os-wireguard-example/

**Comment by chrysn** [comment]
Smart LED support: https://cloudisland.nz/@dinosaur/114579402251842221 / https://blog.mikey.nz/first-look-at-blinksy/

**Reply by chrysn** [reply]
As evidenced by the demo at FrOSCon, we can do smart pixels. @kaspar030, can you share how that worked?

**Comment by chrysn** [comment]
Wall time abstraction.

Having "a clock" as a generic application would be desirable to demonstrate our capability to create write-once-deploy-where-the-shoe-fits applications.

Automatically-selectable backends for wall time abstraction can be:
- any combination of means to get time zones plus any source of UNIX-ish time.
  - We currently don't have an abstraction for UNIXish time, but that could be backed by GPS or NTP (as well as by BLE, see below)
  - Potential sources of time zone updates can be the firmware itself, especially if updates are low-threshold (like, rebuild whenever tzdata releases); might require setting a time zone at build time or with some runtime configuration.
- DCF77 receivers
- BLE (only reasonable when the connected peer provides the [Curren Time Service](https://www.bluetooth.com/specifications/specs/html/?src=CTS_v1.1.0/out/en/index-en.html#UUID-06e3fcba-7cc8-11dd-2dbd-f613fe5cd6bd) including Local Time Information, and the [Next DST Change](https://www.bluetooth.com/specifications/specs/html/?src=NDCS_v1.0/out/en/index-en.html) service, lest people get wrong time readings twice per year until the next sync)

**Comment by Travelbits** [comment]
Future request: Network protocol: DECT 2020/NR+/Mesh 5G. 

ETSI TS 103 636. 

DECT-2020 is the latest development providing very reliable radio transmission supporting advanced channel coding and Hybrid ARQ with incremental redundancy, which enables fast re-transmission. The physical layer supports fast link adaptation, transmit and receiver diversity, as well as MIMO operations up to 8 streams. DECT-2020 covers star network topology, e.g. for ‘Ultra Reliable Low Latency Communications’ (URLLC) use cases, as well as mesh network topology, e.g. for ‘massive Machine Type Communications’ (mMTC) use cases.

**Reply by Travelbits** [reply]
The funding of this request is possible. There is a budget for 4 FTE developers. The project is under huge time stress.

**Reply by chrysn** [reply]
The advertised properties of 1ms hop latency and 6km LoS distance, paired with PHY level HARQ and using rather-quite bands (1.9GHz) definitely make this sound tempting for IoT applications and thus Ariel OS.

**Comment by chrysn** [comment]
At some point, meshtastic support has been asked about at some time in the chat.

That would also be interesting in the sense that services that otherwise use a higher-bandwidth network can be tested on that.

**Comment by SimonIT** [comment]
ZigBee support (maybe with https://github.com/zigbee-rs/zigbee-rs)

**Comment by chrysn** [comment]
For native, it'd be cool to have more coverage of things that can be tested easily:
* GPIO (not necessarily through Linux GPIO, more likely just some host IPC, with some easy management of which GPIOs exist)
* SPI/I2C/UART: could be native (esp. for devices connected through USB UARTs to test hardware w/o MCUs), could be host IPC
* storage ([WIP](https://github.com/lake-wg/authz/issues/74#issuecomment-4061993854))
* USB gadget (e.g. using USB-IP)
* BLE (I don't even know how that'd work, quite possibly by using a USB BLE adapter and forwarding the HCI)
